File: litedb-java/src/main/java/com/litedb/cluster/Pd.java
package com.litedb.cluster;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Pd — the Placement Driver as its OWN Raft group (the TiKV PD model). Mirror of pd.py.
 *
 * The control plane is no longer a single in-process orchestrator (a SPOF). A small odd set of nodes
 * ({@code PD_NODES}) each run a PD Raft replica; membership DECISIONS (add_node / remove_node) are
 * committed to the PD Raft log, so they survive a PD-leader crash. The PD LEADER runs the
 * orchestration: a gossip failure detector (proposes remove_node THROUGH the log) and an idempotent
 * reconcile loop that moves the data cluster toward compute_placement(active), one membership change
 * per shard per pass. Because reconcile is idempotent and derives "current" from the live cluster, a
 * NEW PD leader simply keeps reconciling — a half-finished rebalance is completed, not lost.
 */
public final class Pd {
    private final NodeServer node;
    private final List<String> active;          // replicated membership (state machine)
    private final Map<String, Long> deadSince = new LinkedHashMap<>();
    private final double reapAfterS = 5.0;
    private volatile boolean running = false;
    final RaftGroup raft;

    public Pd(NodeServer node, List<String> pdNodes) {
        this.node = node;
        this.active = new ArrayList<>(ClusterConfig.INITIAL_NODES);
        List<String> peers = new ArrayList<>();
        for (String n : pdNodes) if (!n.equals(node.nodeId())) peers.add(n);
        this.raft = new RaftGroup(node.nodeId(), "pd", peers, node::pdSend, this::apply,
                ClusterConfig.nodeDataDir(node.nodeId()),
                !pdNodes.isEmpty() && pdNodes.get(0).equals(node.nodeId()), node.events, pdNodes);
    }

    public void start() {
        running = true;
        raft.start();
        Thread t = new Thread(this::controlLoop, "pd-control-" + node.nodeId());
        t.setDaemon(true);
        t.start();
    }

    public void stop() {
        running = false;
        raft.stop();
    }

    // ---- replicated state machine ----------------------------------------

    private void apply(long index, Map<String, Object> command) {
        String op = (String) command.get("op");
        if ("add_node".equals(op)) {
            String n = (String) command.get("node");
            if (!active.contains(n)) active.add(n);
        } else if ("remove_node".equals(op)) {
            active.remove((String) command.get("node"));
        }
        // "noop" / "config" carry no placement decision
    }

    // ---- client-facing (leader only) -------------------------------------

    public Map<String, Object> propose(Map<String, Object> decision) {
        Long idx = raft.propose(decision);
        if (idx == null) return NodeServer.resp("ok", false, "leader", raft.leaderId());
        raft.waitCommit(idx, 5000);
        return NodeServer.resp("ok", true);
    }

    public Map<String, Object> status() {
        List<String> a = new ArrayList<>(active);
        return NodeServer.resp("ok", true, "active", a, "placement", ClusterConfig.computePlacement(a),
                "leader", raft.leaderId(), "is_leader", raft.isLeader());
    }

    // ---- PD-leader control loop: failure detection + reconcile -----------

    private void controlLoop() {
        sleep(2000);
        while (running) {
            sleep(1000);
            try {
                if (raft.isLeader()) {
                    detectFailures();
                    reconcile();
                }
            } catch (RuntimeException e) {
                node.events.emit("config", "PD control error: " + e.getMessage());
            }
        }
    }

    @SuppressWarnings("unchecked")
    private void detectFailures() {
        List<String> snapshot = new ArrayList<>(active);
        if (snapshot.size() <= 1) return;
        Map<String, Map<String, Object>> views = new LinkedHashMap<>();
        for (String n : snapshot) {
            Map<String, Object> res = node.callPublic(n, "status", NodeServer.resp());
            if (Boolean.TRUE.equals(res.get("ok")) && res.get("members") instanceof Map) {
                views.put(n, (Map<String, Object>) res.get("members"));
            }
        }
        if (views.isEmpty() || views.size() <= snapshot.size() / 2.0) return;
        long now = System.currentTimeMillis();
        List<String> responders = new ArrayList<>(views.keySet());
        for (String cand : snapshot) {
            int others = 0, deadVotes = 0;
            for (String r : responders) {
                if (r.equals(cand)) continue;
                others++;
                Map<String, Object> e = (Map<String, Object>) views.get(r).get(cand);
                if (e != null && "dead".equals(e.get("state"))) deadVotes++;
            }
            if (others == 0) continue;
            if (deadVotes > others / 2.0) {
                long first = deadSince.computeIfAbsent(cand, k -> now);
                if (now - first >= reapAfterS * 1000) {
                    node.events.emit("config", "PD FAILURE DETECTOR: " + cand + " reported DEAD by "
                            + deadVotes + "/" + others + " peers → proposing remove_node to the PD Raft log");
                    deadSince.remove(cand);
                    Long idx = raft.propose(NodeServer.resp("op", "remove_node", "node", cand, "dead", true));
                    if (idx != null) raft.waitCommit(idx, 5000);
                    return;
                }
            } else {
                deadSince.remove(cand);
            }
        }
    }

    private void reconcile() {
        Map<String, List<String>> desired = ClusterConfig.computePlacement(new ArrayList<>(active));
        for (String shard : ClusterConfig.SHARDS) {
            List<String> want = desired.getOrDefault(shard, List.of());
            String leader = node.leaderOfPublic(shard);
            if (leader == null) continue;
            List<String> cur = votersOf(shard, leader);
            if (cur == null) continue;
            List<String> add = new ArrayList<>();
            for (String n : want) if (!cur.contains(n)) add.add(n);
            List<String> drop = new ArrayList<>();
            for (String n : cur) if (!want.contains(n)) drop.add(n);
            if (!add.isEmpty()) {
                String n = add.get(0);
                node.callPublic(n, "host_shard", NodeServer.resp("shard", shard, "voters", new ArrayList<>(cur)));
                java.util.TreeSet<String> nv = new java.util.TreeSet<>(cur);
                nv.add(n);
                node.callPublic(leader, "reconfigure", NodeServer.resp("shard", shard, "voters", new ArrayList<>(nv)));
            } else if (!drop.isEmpty()) {
                String n = drop.get(0);
                java.util.TreeSet<String> nv = new java.util.TreeSet<>(cur);
                nv.remove(n);
                if (nv.isEmpty()) continue;
                node.callPublic(leader, "reconfigure", NodeServer.resp("shard", shard, "voters", new ArrayList<>(nv)));
                if (alive(n)) node.callPublic(n, "drop_shard", NodeServer.resp("shard", shard));
            }
        }
    }

    @SuppressWarnings("unchecked")
    private List<String> votersOf(String shard, String leader) {
        Map<String, Object> res = node.callPublic(leader, "status", NodeServer.resp());
        if (!Boolean.TRUE.equals(res.get("ok"))) return null;
        for (Object so : (List<Object>) res.getOrDefault("shards", List.of())) {
            Map<String, Object> sh = (Map<String, Object>) so;
            if (shard.equals(sh.get("group"))) {
                List<String> v = new ArrayList<>((List<String>) sh.getOrDefault("voters", List.of()));
                java.util.Collections.sort(v);
                return v;
            }
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    private boolean alive(String nodeId) {
        Object e = node.gossipView().get(nodeId);
        return !(e instanceof Map) || !"dead".equals(((Map<String, Object>) e).get("state"));
    }

    private static void sleep(long ms) {
        try { Thread.sleep(ms); } catch (InterruptedException ignored) {}
    }
}

File: litedb-python/pd.py
"""
pd.py — the Placement Driver as its OWN Raft group (the TiKV PD model).

The control plane used to be a single in-process orchestrator: a SPOF whose death could leave a
rebalance half-applied. Here it is a real **replicated state machine**. A small odd set of nodes
(`PD_NODES`) each run a PD Raft replica; membership *decisions* — `add_node` / `remove_node` — are
committed to the PD Raft log, so they are durable and survive a PD-leader crash. The PD **leader**
runs the orchestration:

  * failure detector — reads gossip liveness; a node confirmed DEAD by a majority of peers past a
    grace window is proposed as `remove_node(dead)` THROUGH the PD log (durable decision).
  * reconcile — an idempotent control loop that moves the data cluster toward
    `compute_placement(active)`, one membership change per shard per pass (add a replica → it catches
    up via Raft → drop the extra). Because it's idempotent and derives "current" from the live
    cluster, a NEW PD leader simply keeps reconciling — a half-finished rebalance is completed, not
    lost. This is the same desired-state-reconciliation a real placement driver does.

So the loop "gossip detected a death → restore RF" now runs inside a fault-tolerant, Raft-replicated
control plane: kill the PD leader and a new one takes over and finishes the job.
"""

import threading
import time

from cluster_config import INITIAL_NODES, SHARDS, compute_placement, node_data_dir
from raft_node import RaftGroup


class Pd:
    def __init__(self, node, pd_nodes: list[str]):
        self.node = node                       # the owning NodeServer (reused for RPC + gossip)
        self.pd_nodes = list(pd_nodes)
        self.active: list[str] = list(INITIAL_NODES)   # replicated membership (state machine)
        self._dead_since: dict[str, float] = {}
        self._running = False
        self._reap_after = 5.0
        self.raft = RaftGroup(
            node_id=node.node_id, group_id="pd",
            peers=[n for n in pd_nodes if n != node.node_id],
            send_fn=node._pd_send, apply_fn=self._apply,
            data_dir=node_data_dir(node.node_id),
            preferred=(bool(pd_nodes) and pd_nodes[0] == node.node_id),
            on_event=node.events.emit, voters=pd_nodes,
        )

    # ---- lifecycle --------------------------------------------------------

    def start(self) -> None:
        self._running = True
        self.raft.start()
        threading.Thread(target=self._control_loop, daemon=True).start()

    def stop(self) -> None:
        self._running = False
        self.raft.stop()

    # ---- replicated state machine ----------------------------------------

    def _apply(self, index: int, command: dict) -> None:
        op = command.get("op")
        if op == "add_node":
            if command["node"] not in self.active:
                self.active.append(command["node"])
        elif op == "remove_node":
            if command["node"] in self.active:
                self.active.remove(command["node"])
        # "noop" / "config" entries carry no placement decision

    # ---- client-facing (leader only) -------------------------------------

    def propose(self, decision: dict) -> dict:
        idx = self.raft.propose(decision)
        if idx is None:
            return {"ok": False, "leader": self.raft.leader_id}
        self.raft.wait_commit(idx, timeout=5.0)
        return {"ok": True}

    def status(self) -> dict:
        active = list(self.active)
        return {"ok": True, "active": active, "placement": compute_placement(active),
                "leader": self.raft.leader_id, "is_leader": self.raft.is_leader()}

    # ---- PD-leader control loop: failure detection + reconcile ------------

    def _control_loop(self) -> None:
        time.sleep(2.0)  # let the PD group elect and the data cluster come up
        while self._running:
            time.sleep(1.0)
            try:
                if self.raft.is_leader():
                    self._detect_failures()
                    self._reconcile()
            except Exception as e:
                self.node.events.emit("config", f"PD control error: {e}")

    def _detect_failures(self) -> None:
        """Read gossip liveness from each live node; a node DEAD by a majority of peers past the
        grace window is proposed as remove_node(dead) THROUGH the PD log (durable decision)."""
        active = list(self.active)
        if len(active) <= 1:
            return
        views: dict[str, dict] = {}
        for n in active:
            res = self.node._call(n, "status", {})
            if res.get("ok") and isinstance(res.get("members"), dict):
                views[n] = res["members"]
        if not views or len(views) <= len(active) / 2:   # need a live majority to act safely
            return
        now = time.time()
        responders = list(views)
        for cand in active:
            others = [r for r in responders if r != cand]
            if not others:
                continue
            dead = sum(1 for r in others if views[r].get(cand, {}).get("state") == "dead")
            if dead > len(others) / 2:
                first = self._dead_since.setdefault(cand, now)
                if now - first >= self._reap_after:
                    self.node.events.emit("config", f"PD FAILURE DETECTOR: {cand} reported DEAD by "
                                          f"{dead}/{len(others)} peers → proposing remove_node to the "
                                          f"PD Raft log (durable decision)")
                    self._dead_since.pop(cand, None)
                    idx = self.raft.propose({"op": "remove_node", "node": cand, "dead": True})
                    if idx is not None:
                        self.raft.wait_commit(idx, timeout=5.0)
                    return
            else:
                self._dead_since.pop(cand, None)

    def _reconcile(self) -> None:
        """Idempotent: move each shard toward compute_placement(active), one membership change at a
        time. 'Current' is observed from the live cluster (the shard leader's Raft config), so a new
        PD leader needs no prior in-memory state — it just keeps reconciling toward the target."""
        desired = compute_placement(self.active)
        for shard in SHARDS:
            want = desired.get(shard, [])
            leader = self.node._leader_of(shard, retries=1)
            if not leader:
                continue
            cur = self._voters_of(shard, leader)
            if cur is None:
                continue
            add = [n for n in want if n not in cur]
            drop = [n for n in cur if n not in want]
            if add:
                n = add[0]
                self.node._call(n, "host_shard", {"shard": shard, "voters": cur})
                self.node._call(leader, "reconfigure",
                                {"shard": shard, "voters": sorted(set(cur) | {n})})
            elif drop:
                n = drop[0]
                new = sorted(set(cur) - {n})
                if not new:
                    continue
                self.node._call(leader, "reconfigure", {"shard": shard, "voters": new})
                if self._alive(n):
                    self.node._call(n, "drop_shard", {"shard": shard})

    def _voters_of(self, shard: str, leader: str):
        res = self.node._call(leader, "status", {})
        if not res.get("ok"):
            return None
        for sh in res.get("shards", []):
            if sh["group"] == shard:
                return sorted(sh.get("voters", []))
        return None

    def _alive(self, node_id: str) -> bool:
        return self.node.gossip.view().get(node_id, {}).get("state") != "dead"
